File: Kode/algoritmer/AVLTree.java
package algoritmer;

import java.util.ArrayList;
import java.util.List;

public class AVLTree {
    public class Node {
        public int x;
        public int height;
        public Node left;
        public Node right;

        Node(int x) {
            this.x = x;
            this.height = 0;
        }
    }

    // første element - roten til treet / starten
    public Node root;



    /** rotasjoner **/
    Node leftRotate(Node z) {
        Node y = z.right;
        Node t1 = y.left;

        y.left = z;
        z.right = t1;

        z.height = 1 + Math.max( height(z.left), height(z.right) );
        y.height = 1 + Math.max( height(y.left), height(y.right) );

        return y;
    }

    Node rightRotate(Node z) {
        Node y = z.left;
        Node t2 = y.right;

        y.right = z;
        z.left = t2;

        z.height = 1 + Math.max( height(z.left), height(z.right) );
        y.height = 1 + Math.max( height(y.left), height(y.right) );

        return y;
    }

    /** balansering **/
    // positivt = venstretung - høyrerotasjon
    // negativt = høyretung - venstrerotasjon
    int balanceFactor(Node v) {
        if (v == null)
            return 0;
        return (height(v.left) - height(v.right));
    }

    Node balance(Node v) {
        // høyretung
        if (balanceFactor(v) < -1) {
            if (balanceFactor(v.right) > 0)
                v.right = rightRotate(v.right);
            return leftRotate(v);
        }

        // venstretung
        if (balanceFactor(v) > 1) {
            if (balanceFactor(v.left) < 0)
                v.left = leftRotate(v.left);
            return rightRotate(v);
        }
        return v;
    }

    /** balance-insert **/// støtter inserting av 1 eller flere
    public void insert(int... args) {
        for (int arg : args)
            root = insert(root, arg);
    }
    Node insert(Node v, int x) {
        if (v == null)
            v =  new Node(x);

        else if (x < v.x)
            v.left = insert(v.left, x);

        else if (x > v.x)
            v.right = insert(v.right, x);

        v.height = 1 + Math.max( height(v.left), height(v.right) );
        return balance(v);
    }


    /** balance-remove **/
    public void remove(int x) {
        root = remove(root, x);
    }
    Node remove(Node v, int x) {
        if (v == null)
            return null;

        if (x < v.x)
            v.left = remove(v.left, x);

        else if (x > v.x)
            v.right = remove(v.right, x);

        // case 2: har kun 1-barnenode
        else if (v.left == null)
            v = v.right;

        else if (v.right == null)
            v = v.left;

        // case 3: har begge barnenoder
        else {
            Node u = findMin(v.right);
            v.x = u.x;
            v.right = remove(v.right, u.x);
            v.height = 1 + Math.max( height(v.left), height(v.right) );
        }
        return balance(v);
    }
    // det står her i forelesningen, men ↑ fungerer, ikke ↓. Spør?
    //v.height = 1 + Math.max( height(v.left), height(v.right) );


    /** contains**/
    public boolean contains(int x) {
        return contains(root, x);
    }
    // sjekker om x finnes i treet
    boolean contains(Node v, int x) {
        if (v == null)
            return false;

        if (x < v.x)
            return contains(v.left, x);

        if (x > v.x)
            return contains(v.right, x);

        return true;
    }

    /** inorder**/
    // inorder = i rekkefølge - minst til størst
    public void inorder() {
        inorder(root);
    }
    void inorder(Node v) {
        if (v == null) return;
        inorder(v.left);
        System.out.print(v.x+"  ");
        inorder(v.right);
    }

    /** postorder **/
    public void postorder() {
        postorder(root);
    }
    void postorder(Node v) {
        if (v == null) return;
        postorder(v.left);
        postorder(v.right);
        System.out.print(v.x+"  ");
    }

    /** preorder **/
    public void preorder() {
        preorder(root);
    }
    void preorder(Node v) {
        if (v == null) return;
        System.out.print(v.x+"  ");
        postorder(v.left);
        postorder(v.right);
    }

    /** hjelpemtoder **/
    Node findMin(Node v) {
        if (v == null || v.left == null) {
            return v;
        }
        return findMin(v.left);
    }

    Node findMax(Node v) {
        if (v == null || v.right == null) {
            return v;
        }
        return findMax(v.right);
    }

    int height(Node v) {
        if (v == null)
            return 0;
        return Math.max(height(v.left), height(v.right)) + 1;
    }

    // broken metode - printer ut et FINT tre https://stackoverflow.com/questions/4965335/how-to-print-binary-tree-diagram-in-java
    public void print() {
        List<List<String>> lines = new ArrayList<List<String>>();
        List<Node> level = new ArrayList<Node>();
        List<Node> next = new ArrayList<Node>();

        level.add(root);
        int nn = 1;

        int widest = 0;

        while (nn != 0) {
            List<String> line = new ArrayList<String>();

            nn = 0;

            for (Node n : level) {
                if (n == null) {
                    line.add(null);

                    next.add(null);
                    next.add(null);
                } else {
                    String aa = Integer.toString(n.x);
                    line.add(aa);
                    if (aa.length() > widest) widest = aa.length();

                    next.add(n.left);
                    next.add(n.right);

                    if (n.left != null) nn++;
                    if (n.right != null) nn++;
                }
            }

            if (widest % 2 == 1) widest++;

            lines.add(line);

            List<Node> tmp = level;
            level = next;
            next = tmp;
            next.clear();
        }

        int perpiece = lines.get(lines.size() - 1).size() * (widest + 4);
        for (int i = 0; i < lines.size(); i++) {
            List<String> line = lines.get(i);
            int hpw = (int) Math.floor(perpiece / 2f) - 1;

            if (i > 0) {
                for (int j = 0; j < line.size(); j++) {

                    // split node
                    char c = ' ';
                    if (j % 2 == 1) {
                        if (line.get(j - 1) != null) {
                            c = (line.get(j) != null) ? '┴' : '┘';
                        } else {
                            if (line.get(j) != null) c = '└';
                        }
                    }
                    System.out.print(c);

                    // lines and spaces
                    if (line.get(j) == null) {
                        for (int k = 0; k < perpiece - 1; k++) {
                            System.out.print(" ");
                        }
                    } else {

                        for (int k = 0; k < hpw; k++) {
                            System.out.print(j % 2 == 0 ? " " : "─");
                        }
                        System.out.print(j % 2 == 0 ? "┌" : "┐");
                        for (int k = 0; k < hpw; k++) {
                            System.out.print(j % 2 == 0 ? "─" : " ");
                        }
                    }
                }
                System.out.println();
            }

            // print line of numbers
            for (String f : line) {

                if (f == null) f = "";
                int gap1 = (int) Math.ceil(perpiece / 2f - f.length() / 2f);
                int gap2 = (int) Math.floor(perpiece / 2f - f.length() / 2f);

                // a number
                for (int k = 0; k < gap1; k++) {
                    System.out.print(" ");
                }
                System.out.print(f);
                for (int k = 0; k < gap2; k++) {
                    System.out.print(" ");
                }
            }
            System.out.println();

            perpiece /= 2;
        }
    }

    public static void main(String[] args) {
        AVLTree t= new AVLTree();

        t.insert(5, 8, 10, 9, 7, 6);
        t.insert(2, 1, 0, -1, 11, 13, 16);

        System.out.println("\n3: "+t.contains(3));
        System.out.println("12: "+t.contains(12));
        System.out.println("7: "+t.contains(7));

        System.out.println("> Fjerner 3 og 7 <");
        t.remove(3);
        t.remove(7);

        System.out.println("3: "+t.contains(3));
        System.out.println("12: "+t.contains(12));
        System.out.println("7: "+t.contains(7));
        System.out.println("---------------------------------");

        System.out.println("Root: "+t.root.x);
        System.out.println("Printer inorder (left-root-right)");
        t.inorder();

        System.out.println("\nPrinter postorder (left-right-root)");
        t.postorder();

        System.out.println("\nPrinter preorder (root-left-right)");
        t.preorder();
        System.out.println("\n----------------------------------------------------------------");

        System.out.println("\nminste: "+t.findMin(t.root).x);
        System.out.println("største: "+t.findMax(t.root).x);
        System.out.println("høyde: "+t.height(t.root));

        System.out.print("\n----------------------------------------------------------------\n");
        t.print();
        System.out.println("----------------------------------------------------------------");
        System.out.print("add: 12 og 3\n");
        t.insert(12, 3);
        t.print();
        System.out.println("\n----------------------------------------------------------------");
        System.out.print("remove: 5 og 11\n");
        t.remove(5);
        t.remove(11);
        t.print();
        System.out.println("----------------------------------------------------------------");

        BinarySearchTree bst = new BinarySearchTree();
        AVLTree avl = new AVLTree();
        bst.insert( 1, 6, 9, 2, 3, -1, -3);
        avl.insert( 1, 6, 9, 2, 3, -1, -3);

        System.out.println("\n\n - Sammenligning av BSt og AVL - ");
        System.out.println("BST - Binary search tree");
        bst.print();
        System.out.println("\nAVL - Balansert BST");
        avl.print();
        System.out.println("\n-----------------------------\n");
    }
}

File: Kode/algoritmer/python_algoritmer/datastrukturer/AVLTree.py
import math

class AVL:
    """AVL-tree class containing subclass and methods"""
    def __init__(self):
        self.root = None

    class Node:
        """Node class containing data and children-pointers"""
        def __init__(self, data):
            self.data = data
            self.height = 0
            self.left = None
            self.right = None

    def leftRotate(self, node):
        """Performs a left-rotate of node and subtree"""
        right_child = node.right
        right_child_left = right_child.left

        right_child.left = node
        node.right = right_child_left

        node.height = 1 + max(self.height(node.left), self.height(node.right))
        right_child.height = 1 + max(self.height(right_child.left), self.height(right_child.right))

        return right_child

    def rightRotate(self, node):
        """Performs a right-rotate of node and subtree"""
        left_child = node.left
        left_child_right = left_child.right

        left_child.right = node
        node.left = left_child_right

        node.height = 1 + max(self.height(node.left), self.height(node.right))
        left_child.height = 1 + max(self.height(left_child.left), self.height(left_child.right))

        return left_child

    def balance(self, node):
        """Balances tree based on balancefactor of current node"""
        # høyretung
        if self.balanceFactor(node) < -1:
            if self.balanceFactor(node.right) > 0:
                node.right = self.rightRotate(node.right)
            return self.leftRotate(node)

        # venstretung
        if self.balanceFactor(node) > 1:
            if self.balanceFactor(node.left) < 0:
                node.left = self.leftRotate(node.left)
            return self.rightRotate(node)

        return node

    # inserting - start
    def insert(self, *args):
        """Inserts element(s) in tree"""
        for arg in args:
            self.root = self.balance_insert(self.root, arg)

    def balance_insert(self, node, element):
        if node is None:
            node = self.Node(element)

        elif element < node.data:
            node.left = self.balance_insert(node.left, element)

        elif element > node.data:
            node.right = self.balance_insert(node.right, element)

        node.height = 1 + max(self.height(node.left), self.height(node.right))
        return self.balance(node)

    # insert - end

    # remove - start
    def remove(self, element):
        """Removes node with given element from tree"""
        self.root = self.balance_remove(self.root, element)

    def balance_remove(self, node, element):
        if node is None:
            return None

        if element < node.data:
            node.left = self.balance_remove(node.left, element)

        elif element > node.data:
            node.right = self.balance_remove(node.right, element)

        # case 2: har kun 1-barnenode
        elif node.left is None:
            node = node.right

        elif node.right is None:
            node = node.left

        # case 3: har begge barnenoder
        else:
            lowest = self.findMin(node.right)
            node.data = lowest.data
            node.height = 1 + max(self.height(node.left), self.height(node.right))

        return self.balance(node)

    # remove - end

    # contains
    def contains(self, element):
        """Looks for element in node"""
        return self.balance_contains(self.root, element)

    def balance_contains(self, node, element):
        if node is None:
            return False

        if element < node.data:
            return self.balance_contains(node.left, element)

        if element > node.data:
            return self.balance_contains(node.right, element)

        return True

    # hjelpemetoder

    def balanceFactor(self, node):
        """Finds balance factor of node
        Used to determined if left-heavy or right-heavy
        """
        if node is None:
            return 0
        return self.height(node.left) - self.height(node.right)

    def height(self, node):
        """Finds height of tree"""
        if node is None:
            return 0
        return max(self.height(node.left), self.height(node.right)) + 1

    def findMin(self, node):
        """Finds the lowest node by traversing to left-children"""
        if node is None or node.left is None:
            return node
        return self.findMin(node.left)

    def findMax(self, node):
        """Finds the highest node by traversing to right-children"""
        if node is None or node.right is None:
            return node
        return self.findMax(node.right)

    # traversering

    # inorder - start
    def inorder(self):
        """Prints the AVL-tree inorder - Left-Root-Right of tree"""
        self.print_inorder(self.root)

    def print_inorder(self, node):
        if node is None: return
        self.print_inorder(node.left)
        if node.data == self.root.data: print("[", node.data, "]", end=" ")
        else: print(node.data, end=" ")
        self.print_inorder(node.right)
    # inorder - end

    # postorder - start
    def postorder(self):
        """Prints the AVL-tree postorder - Left-Right-Root of tree"""
        self.print_postorder(self.root)

    def print_postorder(self, node):
        if node is None: return
        self.print_postorder(node.left)
        self.print_postorder(node.right)
        if node.data == self.root.data: print("[", node.data, "]", end=" ")
        else: print(node.data, end=" ")
    # postorder - end

    # preorder - start
    def preorder(self):
        """Prints the AVL-tree preorder - Root-Left-Right of tree"""
        self.print_preorder(self.root)

    def print_preorder(self, node):
        if node is None: return
        if node.data == self.root.data: print("[", node.data, "]", end=" ")
        else: print(node.data, end=" ")
        self.print_preorder(node.left)
        self.print_preorder(node.right)
    # preorder - end

    def print(self):
        print("\n")
        lines = []
        level = []
        next = []

        level.append(self.root)
        nn = 1
        widest = 0

        while nn != 0:
            line = []

            nn = 0

            for node in level:
                if node is None:
                    line.append(None)
                    next.append(None)
                    next.append(None)
                else:
                    aa = str(node.data)
                    line.append(aa)
                    if len(aa) > widest: widest = len(aa)

                    next.append(node.left)
                    next.append(node.right)

                    if node.left is not None: nn += 1
                    if node.right is not None: nn += 1

            if widest % 2 == 1: widest += 1

            lines.append(line)

            tmp = level
            level = next
            next = tmp
            next.clear()

        perpiece: int = len(lines[len(lines) - 1]) * (widest + 4)
        for i in range(len(lines)):
            line = lines[i]
            hpw: int = math.floor(perpiece / 2.0) - 1

            if i > 0:
                for j in range(len(line)):
                    c = ' '
                    if j % 2 == 1:
                        if line[j - 1] is not None:
                            c = '┴' if line[j] is not None else '┘'
                        else:
                            if line[j] is not None:
                                c = '└'
                    print(c, end="")

                    if line[j] is None:
                        for k in range(len(perpiece) - 1):
                            print(" ", end="")
                    else:
                        for k in range(hpw):
                            print(" " if j % 2 == 0 else "─", end="")
                        print("┌" if j % 2 == 0 else "┐", end="")
                        for k in range(hpw):
                            print("─" if j % 2 == 0 else " ", end="")

                print()

            for f in line:
                if f is None: f = ""
                gap1: int = math.ceil(perpiece / 2.0 - len(f) / 2.0)
                gap2: int = math.floor(perpiece / 2.0 - len(f) / 2.0)

                for k in range(gap1):
                    print(" ", end="")
                print(f, end="")
                for k in range(gap2):
                    print(" ", end="")

            print()
            perpiece /= 2



# printing
t = AVL()
t.insert(1, 6, 9, 2, 3, -1, -3)

print("3: ", t.contains(3))
print("12: ", t.contains(12))
print("7: ", t.contains(7))

print("\nRoot of tree:", t.root.data)
print("\n-- Printing inorder --")
t.inorder()
print("\n\n-- Printing postorder  --")
t.postorder()
print("\n\n-- Printing preorder --")
t.preorder()
t.print()
